### PR TITLE
ISSUE-54: makes all blocking calls Promises that can async/await

### LINKED ISSUE #54: ISSUE-54: Make all blocking calls Promises that can async/await
# Description

Many of the low level calls must do DNS lookups or CPU intensive work.  These actions will block the node.js event loop.  For this reason, we should lean on lower level threads handling the work of the addon, while allowing node.js's event loop to move on while `libopendkim` is working.

Since this code is already being used in plugins like `haraka-plugin-opendkim` in an imperative calling style, I think it would be best to make the interface return Promises.  By doing so, we allow the caller to use async/await and significantly reduce the work needed to conform to a callback interface.

# PR must contain

- [ ] a first commit title that starts with 'ISSUE-N:'
- [ ] code
- [ ] tests
- [ ] documentation

# Example PR
https://github.com/godsflaw/node-opendkim/pull/50

# More Information
http://www.opendkim.org/libopendkim/

# Bounty
None

### PR DESCRIPTION
**Pull request by godsflaw**
# Description

Many of the low level calls must do DNS lookups or CPU intensive work.  These actions will block the node.js event loop.  For this reason, we should lean on lower level threads handling the work of the addon, while allowing node.js's event loop to move on while `libopendkim` is working.

Since this code is already being used in plugins like `haraka-plugin-opendkim` in an imperative calling style, I think it would be best to make the interface return Promises.  By doing so, we allow the caller to use async/await and significantly reduce the work needed to conform to a callback interface.

# Contribution Checklist

- [x] first commit title starts with 'ISSUE-N:'
- [x] code approved
- [x] tests approved
- [x] documentation approved
- [x] fixes #54 

### Link to gist documentation (will be manually added to the wiki)
https://gist.github.com/...

### CODE REVIEW

**Review by chazomaticus — DISMISSED** [review]
I didn't look super closely at the tests or try to run it myself, but code-wise this looks good to me!

One thing that popped out is that index.js could be shortened with a `promisify` function to handle the work of creating promises and fulfilling them at the right time.  There also might be an easy like `check_args` function that could also save a little space, but no big deal in either case.

All the new async .cc and .h files seem pretty straightforward.  I wonder if you could shorten some of them by implementing a default HandleOKCallback in the base worker.  Either way, that's only a few lines of code, so again not a big deal.

It's kind of unfortunate to support both sync and async versions of everything, but it does seem like a good idea to have both, since the library is all synchronous, and it's hard to call async code synchronously without muddying up the code significantly.

In summary, there might be a few little ways to keep the code concise, but I heartily approve either way.  Thanks!

**Comment by godsflaw** [comment]
@chazomaticus thanks for the review.  Looks like I need to async one more function `opendkim.flush_cache()` since it could use an underlying DB, and thus could block.  I also still have to solve the unwanted garbage collection problem (at least that is what I think it is).

**Review by chazomaticus — APPROVED** [review]
LGTM!
